server/src/main/java/org/elasticsearch/cluster/metadata/TemplateUpgradeService.java

```diff
@@ -175,27 +172,51 @@ public void onFailure(Exception e) {
             client.admin().indices().deleteTemplate(request, new ActionListener<DeleteIndexTemplateResponse>() {
                 @Override
                 public void onResponse(DeleteIndexTemplateResponse response) {
-                    updatesInProgress.decrementAndGet();
                     if (response.isAcknowledged() == false) {
+                        allUpgradesSuccessful = false;
                         logger.warn("Error deleting template [{}], request was not acknowledged", template);
                     }
+                    tryFinishUpdate();
                 }
 
                 @Override
                 public void onFailure(Exception e) {
-                    updatesInProgress.decrementAndGet();
+                    allUpgradesSuccessful = false;
                     if (e instanceof IndexTemplateMissingException == false) {
                         // we might attempt to delete the same template from different nodes - so that's ok if template doesn't exist
                         // otherwise we need to warn
                         logger.warn(new ParameterizedMessage("Error deleting template [{}]", template), e);
                     }
+                    tryFinishUpdate();
                 }
             });
         }
     }
 
-    int getUpdatesInProgress() {
-        return updatesInProgress.get();
+    private void tryFinishUpdate() {
+        assert upgradesInProgress.get() > 0;
+        if (upgradesInProgress.decrementAndGet() > 1) {
```

Comment:
how about changing this to `if (upgradesInProgress.decrementAndGet() == 1) {` so we can remove the return statement. I find this easier to read as the action only happens when the in progress value is 1.

Reply:
Done! Yes, I have developed a habit of early return conditions no matter the circumstances.